### DISCUSSION TITLE
Ionide says the C# extension is optional, but requires it anyway

### DISCUSSION THREAD

**Discussion by ghost** [ionide/ionide-vscode-fsharp#1470] [General]
I for one never use the debugging support provided, and it's a non-trivial (~150 MiB) amount of stuff to download. The extension should just provide FSAC and other F# features.

The C# extension downloads a bunch of potentially unnecessary support as well, like ASP.NET Core and Razor support which I also never use. To make things worse, ionide doesn't even start working until everything is installed. So I have to wait for C# features I'll never use to be installed before I can get to editing. Overall, this makes the F# extension feel needlessly bloated.

I am fully in favor of incremental complexity. Start with a base F# extension, and let the user add features they need (Debugging, C# support, etc) as needed. The extension should make this adding of features easy, or maybe an extension pack can provide this extension grouping for people that want all of the things right away.

**Comment by mohaalak** [comment]
i think you are absolutely right.

**Reply by dlidstrom** [reply]
I rarely use debugging either, but I find it invaluable the times I do.

**Reply by ghost** [reply]
> I rarely use debugging either, but I find it invaluable the times I do.

Which is why the recommendation I made is:

> Start with a base F# extension, and let the user add features they need (Debugging, C# support, etc) as needed. The extension should make this adding of features easy, or maybe an extension pack can provide this extension grouping for people that want all of the things right away.

**Reply by cartermp** [reply]
Note that C# support and Debugging are one and the same right now. There is no way to get a debugger without also getting the C# extension.

**Reply by baronfel** [reply]
_sorta_. For the OpenVSIX build of ionide we bundle Samsung's debugger, which implements the CoreCLR debugging spec as well.

**Comment by lukemcdo** [comment]
I'd be interested in seeing the Fable REPL adapted to provide an F# extension for VSCode. Perhaps an Ionide-Fable extension? The upside for this approach is an Ionide-Fable extension would be a good fit for running in the new browser-only VSCode modes, as well as faster/easier install in places like GitPod/Code Spaces. 

Point is, that might need to be a separate plugin from this plugin.

**Reply by baronfel** [reply]
Fable Repl isn't viable for .net code, they're two entirely different debugging systems. For Fable code that you want to debug you can use any of the VSCode debuggers that support javascript, from the `node` debug type for nodejs apps to browser-based dubuggers like firefox/chrome.  All of these are entirely incapable of supporting .Net code, by design. 

The changes required to make a browser-only version of Ionide go beyond debugging. The extension depends heavily on msbuild, as well as the entire .net sdk, to power project cracking and tooling analysis, so that would need to be backfilled. This is a _huge_ task.  At minimum, from what I understand of it, web filesystem APIs would need to be opened up to allow access to locations outside of the 'current workspace'. I encourage explorations and contributions into this area, but speaking from experience here - project cracking is a giant endless pit of pain and misery.

**Reply by lukemcdo** [reply]
Obviously the Fable REPL isn't viable for debugging .NET code, but it's viable for syntax highlighting in an environment that lacks a .NET runtime and could provide intellisense type signatures. The original post mentions just wanting FSAC and other F# features. The question from there is just whether FSAC can run on Fable. It's clear from the Fable REPL that something parallel to FSAC could run on Fable. That to me suggests this issue could be at least partially resolved via an Ionide-Fable extension.

I understand that the project file work would be quite hard. I do think though that there would be a lot of interest in the project I'm describing. If you disagree though, you'd probably know better than I.

**Reply by baronfel** [reply]
The good thing about FCS/FSAC is that they are in fact transpilable, yeah :) I would be super in favor of that, but Fable REPL has it easy specifically because they shortcut the project cracking aspect of things - there's no arbitrary project cracking at all, only a blessed set of libraries.  So if that part is solvable, then a road becomes more clear!

**Reply by lukemcdo** [reply]
Thank you so much for explaining this. I might start asking around on that front and try to understand more.

**Reply by baronfel** [reply]
Adding a bit more context: almost every useful FCS API needs the `Project Options`, which is a structure that captures all of the inputs to the compiler for a particular dll output. This includes, but isn't limited to:

* dll references from nuget, explicit references, frameworkreferences, etc
* compiler defines
* target profile (netcore/mscorlib/etc)
* warning settings
* the list of source files
* and many more exciting things

this structure is then used with individual files or whole lists of files to create parse results and check results, which contain most of the useful data.  there's then a layer of 'service' modules that provide specialized, codified access to things.

Project cracking is a term that us oldies will use to refer to converting a fsproj file (or any *proj file, really) into an instance of the Project Options. It sucked before the dotnet sdk because things weren't quite so regularized, and it sucks now because we have this explicit dependency on the SDK.  There are other alternatives here as well, for example build systems like bazel could sidestep the need for a project file if they had some component that could take bazel build files and create a Project Options structure from them.

To make a purely-nodejs hosted set of tooling then, for most .net code you'd need to also transpile or somehow run the SDK, or implement your parallel and probably incorrect version of cracking. Once you do that, however, you then run into the next hurdle: it's likely parts of FCS itself are not fable-friendly. Specifically what I'm thinking of is anywhere that memory-mapped I/O has been used.  So it would likely be a process of discovery to get to a place where FCS itself would be hostable in Fable in its entirety.

**Reply by lukemcdo** [reply]
I can't tell if I'm misunderstanding the situation but it looks to me like FCS gave way to to whatever is in the dotnet/fsharp repository. If that's still colloquially called FCS then that's good to know. Fable appears to simply build that dotnet/fsharp repository with Fable for its compiler services package, though it could be with things manually excluded I suppose.

I can ask more about Fable's exact situation in their issues/forums. 

To be clear though, the objective of a theoretical "Ionide-Fable" would be to provide F# language services alongside Fable-specific tooling, not to replace vscode-fsharp. The original poster isn't asking for debugging support or deep .NET analysis, they're asking for a simple and light install. Providing an Ionide-FSharp-NoDeps for this sounds annoying and confusing. An Ionide-Fable would not be all that much heavier than a theoretical Ionide-FSharp-NoDeps and it'd meet this specific need (along with a myriad of other benefits).

**Reply by Krzysztof-Cieslak** [reply]
This myriad of benefits would include 99% of editor features not working, I guess.

**Reply by cartermp** [reply]
> An Ionide-Fable would not be all that much heavier than a theoretical Ionide-FSharp-NoDeps and it'd meet this specific need (along with a myriad of other benefits).

Not at all. It would have to re-create everything that all dependencies provide today. These dependencies provide the _data_ that Ionide and FCS features use to power tooling. Without that data there is no tooling.

**Reply by lukemcdo** [reply]
> 
> 
> This myriad of benefits would include 99% of editor features not working, I guess.

Thank you for your cynicism, it was extremely constructive. 

> 
> 
> > An Ionide-Fable would not be all that much heavier than a theoretical Ionide-FSharp-NoDeps and it'd meet this specific need (along with a myriad of other benefits).
> 
> Not at all. It would have to re-create everything that all dependencies provide today. These dependencies provide the _data_ that Ionide and FCS features use to power tooling. Without that data there is no tooling.

No, it would have a subset of features. It would not provide debugging support for F# apps targeting .NET Core/5+, for example. That would cut the need for OmniSharp as well as .NET Core debugging support, and for some users it could mean that the .NET SDK isn't needed at all. 

I'll open an issue for more discussion since we're now off-topic from the original post, but from what it sounds like to me, an Ionide-VSCode-Fable could resolve this issue in addition to providing further value to Fable users.

**Reply by cartermp** [reply]
> No, it would have a subset of features. 

Technically, yes that's true. Empty completion lists, no cross-project support of any kind, no ability to detect .NET dependencies (e.g., no understanding of what `Console` is), the list goes on. You might get some basic syntax coloring, but not semantic coloring.

I don't know what you're trying to argue here, frankly. It's an interesting proposition but it's a very expensive one.

**Reply by Krzysztof-Cieslak** [reply]
>  .NET SDK isn't needed at all.

I do think something called a ".NET **software development** kit" is a very reasonable requirement for editor tooling used to **develop software** for a language running on .NET. 

> Thank you for your cynicism, it was extremely constructive.

It's more constructive than this whole discussion so far. People in this thread are amongst the most experienced F#-editor-tooling-developers in the world, and you are completely ignoring any arguments they make. Obviously, you don't have enough context on how F# editor tooling works (which is completely fine by itself - I definitely didn't know how stuff works when I was starting), continuing ignoring arguments given by the people who have this context looks like bad faith behaviour. So I wouldn't recommend continuing it.

**Reply by lukemcdo** [reply]
> 
> 
> > No, it would have a subset of features.
> 
> Technically, yes that's true. Empty completion lists, no cross-project support of any kind, no ability to detect .NET dependencies (e.g., no understanding of what `Console` is), the list goes on. You might get some basic syntax coloring, but not semantic coloring.
> 
> I don't know what you're trying to argue here, frankly. It's an interesting proposition but it's a very expensive one.

I understand your skepticism but the semantic coloring and completion list features are already present at [the Fable repl demo site](https://fable.io/repl/). Obviously both are limited in scope compared to the VSCode completions since the project-related features and .NET dependencies aren't present. I'm sure though that more is possible.

While I can understand how what I'm suggesting might not appear worth its tradeoffs, I feel like I'm seeing a lot of people looking for their tooling to appear on the fancy new vscode.dev site. I'm also seeing more and more projects using either GitPod or GitHub Codespaces. The VSCode site will likely never support Ionide-VSCode-FSharp as-is. GitPod supports it, but it's ugly to set up. I haven't tried setting it up on Codespaces, but I doubt the simplicity would match a one-click install of a proposed Ionide-VSCode-Fable. 

I can understand if you don't prioritize these environments, but I think they'll be important in the coming years.

@Krzysztof-Cieslak I'm happily listening to feedback. I've got the message that this would be damn hard, and I've been asking further questions as to how it would be hard and why, and where I should investigate and try to learn. Am I supposed to act completely overwhelmed by this information that a hard problem is hard?

**Reply by cartermp** [reply]
> I understand your skepticism but the semantic coloring and completion list features are already present at the Fable repl demo site. Obviously both are limited in scope compared to the VSCode completions since the project-related features and .NET dependencies aren't present. I'm sure though that more is possible.

Yes, that's sort of my point. What the fable REPL relies on is roughly just being a standalone F# script file. This is awesome for learning F#, but nowhere near enough to support things like:

* Scripts that interact with your development environment (e.g., file system access)
* Full-on fable projects, since these require a project file and .NET to build
* .NET projects, which the majority of F# developers use today
* Solutions with multiple projects, including mixed C# and F# (which is also fairly common)

If you're in this environment, what powers the Fable REPL won't give you much of anything. You need components like what the .NET SDK give you to generate the data needed to make the plugin have value in any of these scenarios. Replacing this with Fable means you'd need to re-create _all of that_, which is extremely challenging and error-prone work. There's a reason why there's entire software engineering teams at Microsoft who build this stuff that we sit on top of.

> I'm also seeing more and more projects using either GitPod or GitHub Codespaces. The VSCode site will likely never support Ionide-VSCode-FSharp as-is.

Gitpod supports .NET today (IIRC they're having a kernel bug right now but that's just a bug). It's just a container image with the SDK on it. I've used it for a year now and Ionide works great with it.

The vscode site is an interesting one to consider, since it has restricted capabilities. Likely just supporting F# scripts via Fable is possible. If there were a way to let plugins know which host is loading them then that may be possible.

**Reply by baronfel** [reply]
> IIRC they're having a kernel bug right now but that's just a bug

Oh thank goodness, I thought it was something I'd done that was wrong on my testing gitpod workspaces.

**Reply by lukemcdo** [reply]
> 
> 
> > I understand your skepticism but the semantic coloring and completion list features are already present at the Fable repl demo site. Obviously both are limited in scope compared to the VSCode completions since the project-related features and .NET dependencies aren't present. I'm sure though that more is possible.
> 
> Yes, that's sort of my point. What the fable REPL relies on is roughly just being a standalone F# script file. This is awesome for learning F#, but nowhere near enough to support things like: ...

I understood that just fine from prior messages from @/baronfel. I've been brainstorming possibilities such as grabbing project parsing from Mono and seeing if that will run in a WASM environment as an interim solution.

**Reply by cartermp** [reply]
It's more than just parsing projects though. You'll need to actually know which assemblies to load so that the F# compiler service knows which types to bind to so that it can offer accurate typechecking. Failing that, this capability (distribution?) of the extension won't support F# targeting .NET. We're talking something on the order of years to make it robust enough to support what people expect.
